openshift/jenkins/Jenkinsfile

```diff
@@ -0,0 +1,323 @@
+              def feBase = openshift.selector("bc", build_config)
+              if (!feBase.exists()) {
+                throw new Exception("Could not find FRONTEND-base build with name: '${build_config}'")
+              } else {
+                timeout(WAIT_TIMEOUT) {
+                  feBase.startBuild("-w")
+                }
+              }
+
+              // Build the runtime image, if you are not using an off the shelf one.
+              def runtime_build_config = "${APP_NAME}-${COMP_NAME}-${DEPLOY_TO}"
+              def feRuntime = openshift.selector("bc", runtime_build_config)
+              if (!feRuntime.exists()) {
+                throw new Exception("Could not find FRONTEND-runtime build with name: '${runtime_build_config}'")
+              } else {
+                timeout(WAIT_TIMEOUT) {
+                  feRuntime.startBuild("-w")
+                }
+              }
+            }
+          }
+        }
+      }
+      post {
+        success {
+          script {
+            env.FE_BUILT = "true"
+          }
+        }
+      }
+    }
+    stage("Run Tests") {
+      agent any
+      when {
+        anyOf {
+          environment name: "API_BUILT", value: "true"
+          environment name: "FE_BUILT", value: "true"
+        }
+      }
+      steps {
+        echo "[ci/cd]  Tests should be run here"
+      }
+    }
+    stage("Deploy API to DEV") {
+      agent any
+      environment {
+        COMP_NAME = "api"
+      }
+      when { environment name: "API_BUILT", value: "true" }
+      steps {
+        script {
+          openshift.withCluster() {
+            openshift.withProject(TOOLS_PROJECT) {
+              def image_name = "${APP_NAME}-${COMP_NAME}"
+              // Don't tag with BUILD_ID so the pruner can do it's job; it won't delete tagged images.
+              // Tag the images for deployment based on the image's hash
+              def hash = cli.getImageTagHash("${image_name}")
+              echo "[ci/cd]  Tagging ${image_name}@${hash} for deployment to ${DEPLOY_TO} ..."
+              openshift.tag("${image_name}@${hash}", "${image_name}:${DEPLOY_TO}")
+            }
+
+            openshift.withProject(DEV_PROJECT) {
+                def dc = openshift.selector("dc", [ app: APP_NAME, component: COMP_NAME ])
+                echo "[ci/cd]  Found ${dc.count()} DeploymentConfigs: ${dc.names()}"
+                // Wait for the deployment to complete.
+                // This will wait until the desired replicas are all available
+                dc.rollout().status()
+            }
+          }
+        }
+        echo "[ci/cd]  Deployment done!"
+      }
+      post {
+        success {
+          script {
+            env.API_DEPLOYED = "true"
+          }
+        }
+      }
+    }
+    stage("Deploy FRONTEND to DEV") {
+      agent any
+      environment {
+        COMP_NAME = "app"
+      }
+      when { environment name: "FE_BUILT", value: "true" }
```

Comment:
This should handle the force build option.  Also should check if the API deployed successfully.

Reply:
Fixed